File: kickstart/2017/Round-B/A.rb
T = gets.chomp.to_i

X = 10**9 + 7
MAX_K = 10000 + 1
mpow = Array.new(MAX_K, 0)
mpow[0] = 1
(1..MAX_K).each do |i|
  mpow[i] = mpow[i-1] * 2 % X
end

T.times do |c|
  ret = 0
  n = gets.chomp.to_i
  k = gets.chomp.split(" ").map(&:to_i)
  n.times do |i|
    ((i+1)..(n-1)).each do |j|
      # curr = 2**(j - i - 1)
      curr = mpow[j - i - 1]
      ret += ((k[j] - k[i]) * curr) % X
      ret %= X
    end
  end
  puts "Case ##{c+1}: #{ret % X}"
end

File: kickstart/2017/Round-B/A.cpp
#include <iostream>
#include <algorithm>

using namespace std;

const int X = 1000000000 + 7;
const int MAX_K = 10000 + 1;

int T,
    N;
int K[MAX_K];
int mpow[MAX_K];
long long curr;

int main()
{
  mpow[0] = 1;
  for (int i = 1; i < MAX_K; i++)
  {
    mpow[i] = mpow[i - 1] * 2 % X;
  }
  ios::sync_with_stdio(false);
  cin >> T;
  for (int c = 0; c < T; c++)
  {
    long long ret = 0;
    cin >> N;
    for (int i = 0; i < N; i++)
    {
      cin >> K[i];
    }
    for (int i = 0; i < N; i++)
    {
      for (int j = i + 1; j < N; j++)
      {
        curr = mpow[j - i - 1];
        ret += ((K[j] - K[i]) * curr) % X;
        ret %= X;
      }
    }
    cout << "Case #" << c + 1 << ": " << ret % X << endl;
  }
  return 0;
}
